### PR TITLE
Increase handler test coverage

### PR DESCRIPTION
**Pull request by simone-kalbermatter**
This PR addresses issue [#1378](https://github.com/dedis/popstellar/issues/1378) and increases the test coverage of `LaoHandler` and `ElectionHandler`. Because I needed `PendingUpdate` in my tests, I also added an `equals` and `hashCode` method there, and created the corresponding test class `PendingUpdateTest`.

### CODE REVIEW

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/LaoHandler.java [hunk 1]
```diff
@@ -159,15 +161,14 @@ public void handleStateLao(HandlerContext context, StateLao stateLao)
...
-
     lao.setId(stateLao.getId());
     lao.setWitnesses(stateLao.getWitnesses());
     lao.setName(stateLao.getName());
     lao.setLastModified(stateLao.getLastModified());
     lao.setModificationId(stateLao.getModificationId());
 
     PublicKey publicKey = keyManager.getMainPublicKey();
-    if (laoView.isOrganizer(publicKey) || laoView.isWitness(publicKey)) {
+    if (lao.getOrganizer().equals(publicKey) || lao.getWitnesses().contains(publicKey)) {
```
Comment (matteosz) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/LaoHandler.java:171]: Why did you change the previous condition `laoView.isOrganizer(publicKey) || laoView.isWitness(publicKey)`, isn't it better to use such methods so that it's cleaner?

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/model/objects/PendingUpdateTest.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+    assertNotEquals(PENDING_UPDATE1, PENDING_UPDATE4);
+    assertNotEquals(PENDING_UPDATE1.hashCode(), PENDING_UPDATE4.hashCode());
+  }
+
+  @Test
+  public void testCompareTo() {
+    assertTrue(PENDING_UPDATE1.compareTo(PENDING_UPDATE3) < 0);
+    assertTrue(PENDING_UPDATE3.compareTo(PENDING_UPDATE1) > 0);
+    assertTrue(PENDING_UPDATE1.compareTo(PENDING_UPDATE2) == 0);
```
Comment (matteosz) [inline fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/model/objects/PendingUpdateTest.java:53]: Could it better to use `assertEquals` here?

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java [hunk 3]
```diff
@@ -39,7 +41,7 @@ public ElectionHandler(
   /**
    * Process an ElectionSetup message.
    *
-   * @param context the HandlerContext of the message
+   * @param context       the HandlerContext of the message
```
Comment (matteosz) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java:44]: There's an empty tab space here

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java [hunk 4]
```diff
@@ -85,19 +87,18 @@ public void handleElectionSetup(HandlerContext context, ElectionSetup electionSe
   /**
    * Process an ElectionResult message.
    *
-   * @param context the HandlerContext of the message
+   * @param context        the HandlerContext of the message
```
Comment (matteosz) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java:90]: Also here

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java [hunk 5]
```diff
@@ -125,7 +126,7 @@ private Map<String, Set<QuestionResult>> computeResults(
   /**
    * Process an OpenElection message.
    *
-   * @param context the HandlerContext of the message
+   * @param context      the HandlerContext of the message
```
Comment (matteosz) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java:129]: Also here

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java [hunk 6]
```diff
@@ -153,7 +154,7 @@ public void handleElectionOpen(HandlerContext context, OpenElection openElection
   /**
    * Process an ElectionEnd message.
    *
-   * @param context the HandlerContext of the message
+   * @param context     the HandlerContext of the message
```
Comment (matteosz) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java:157]: Also here

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java [hunk 7]
```diff
@@ -247,7 +248,7 @@ public static WitnessMessage electionSetupWitnessMessage(MessageID messageId, El
   /**
    * Simple way to handle a election key, add the given key to the given election
    *
-   * @param context context
+   * @param context     context
```
Comment (matteosz) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java:251]: And here

**Review by matteosz — CHANGES_REQUESTED** [review]
Good work! Just few suggestions

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java (hunk 3 above)
Comment (simone-kalbermatter) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java:44]: I see that my auto format does this for some reason to align it with the parameter of the next line😕 . I'll revert and disable it.

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/model/objects/PendingUpdateTest.java (hunk 2 above)
Comment (simone-kalbermatter) [reply fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/model/objects/PendingUpdateTest.java:53]: For the last one yes, for the first two no because it's a comparison.

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/LaoHandler.java (hunk 1 above)
Comment (simone-kalbermatter) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/LaoHandler.java:171]: I did this to make it consistent with the same line in `handleCreateLao` (line 67), but you are right, the other code is cleaner. I will revert this and instead change the line in `handleCreateLao` to use the `LaoView` too.

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java (hunk 3 above)
Comment (simone-kalbermatter) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java]: There were some other formatting issues too because I was using another laptop and forgot to tick enable google-java-format 🙃. But it's fixed now.

**Review by matteosz — DISMISSED** [review]
LGTM

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java [hunk 8]
```diff
@@ -200,11 +208,20 @@ private void handleElectionResults(Set<QuestionResult> results, Channel election
     messageHandler.handleMessage(messageSender, electionChannel, message);
   }
 
+  private void setMessageRepoReturns(MessageGeneral message) {
+    messageRepo = mock(MessageRepository.class);
```
Comment (Warhedgehog) [inline fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java:212]: @GabrielFleischer do you think it would be better to use Hilt mock injection ?

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java [hunk 9]
```diff
@@ -287,22 +338,114 @@ public void testHandleElectionEnd()
...
     // The expected hash is made on the sorted vote ids
-    String[] voteIds = Stream.of(vote1, vote2).map(Vote::getId).sorted().toArray(String[]::new);
+    String[] voteIds = Stream.of(VOTE1, VOTE2).map(Vote::getId).sorted().toArray(String[]::new);
+    Election election = electionRepo.getElectionByChannel(OPEN_BALLOT_ELECTION.getChannel());
+
+    assertEquals(Hash.hash(voteIds), election.computeRegisteredVotesHash());
+  }
+
+  @Test
+  public void castVoteOnlyKeepsLastVote()
```
Comment (Warhedgehog) [inline fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java:357]: Very good test !

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java [hunk 10]
```diff
@@ -287,22 +338,114 @@ public void testHandleElectionEnd()
...
+    setMessageRepoReturns(nullData);
+
+    handleCastVote(VOTE1, SENDER_KEY, OPENED_AT);
+    assertThrows(
+        IllegalStateException.class, () -> handleCastVote(VOTE1, SENDER_KEY, OPENED_AT + 1));
+  }
+
+  @Test
+  public void castVoteFailsOnPreviousDataInvalid()
+      throws UnknownElectionException, UnknownRollCallException, UnknownLaoException,
```
Comment (Warhedgehog) [inline fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java:416]: I think this test function deserves a comment explaining what you test. It is not immediately clear from the code and the function name IMO

**Review by Warhedgehog — CHANGES_REQUESTED** [review]
Just details including a question for Gabriel. Very good work, well documented ! Great to finally have some tests for cast vote handlers

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java (hunk 8 above)
Comment (GabrielFleischer) [reply fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java:212]: It's true that this is not perfect.

I don't think you need a mock here, you could simply call 
`messageRepo.addMessage(message);`

As the message repository is a simple class, mocking it comes down to reimplementing it.

On line 111, you already added an instantiation of the base class, you should simply use it 😄

**Review by jbsv — APPROVED** [review]
LGTM, nice work

**Review by Warhedgehog — APPROVED** [review]
LGTM
